spaceship/lib/spaceship/api/client.rb

```diff
@@ -0,0 +1,141 @@
+require_relative '../client'
+
+module Spaceship
+  module API
+    class Client < Spaceship::Client
+      ##
+      # Spaceship HTTP client for the App Store Connect API.
+      #
+      # This client is solely responsible for the making HTTP requests and
+      # parsing their responses. Parameters should be either named parameters, or
+      # for large request data bodies, pass in anything that can resond to
+      # `to_json`.
+      #
+      # Each request method should validate the required parameters. A required parameter is one that would result in 400-range response if it is not supplied.
+      # Each request method should make only one request. For more high-level logic, put code in the data models.
+
+      def self.hostname
+        'https://appstoreconnect.apple.com/iris/v1/'
+      end
+
+      def build_url(path: nil, filter: nil, includes: nil, limit: nil, sort: nil)
+        filters = filter.keys.map do |key|
+          "&filter[#{key}]=#{filter[key]}"
+        end.join("")
+        return "#{path}?&include=#{includes}&limit=#{limit}&sort=#{sort}#{filters}"
+      end
+
+      def get_builds(filter: {}, includes: "buildBetaDetail,betaBuildMetrics", limit: 10, sort: "uploadedDate")
+        assert_required_params(__method__, binding)
+
+        # GET
+        # https://appstoreconnect.apple.com/iris/v1/builds
+        url = build_url(path: "builds", filter: filter, includes: includes, limit: limit, sort: sort)
+
+        response = request(:get, url)
+        handle_response(response)
+      end
+
+      def post_for_testflight_review(build_id: nil)
+        assert_required_params(__method__, binding)
+
+        # POST
+        # https://appstoreconnect.apple.com/iris/v1/betaAppReviewSubmissions
+        #
+        # {"data":{"type":"betaAppReviewSubmissions","relationships":{"build":{"data":{"type":"builds","id":"6f90f04a-3b48-4d4a-9bbd-9475c294a579"}}}}}
+
+        body = {
+          data: {
+            type: "betaAppReviewSubmissions",
+            relationships: {
+              build: {
+                data: {
+                  type: "builds",
+                  id: build_id
+                }
+              }
+            }
+          }
+        }
+
+        response = request(:post) do |req|
+          req.url("betaAppReviewSubmissions")
+          req.body = body.to_json
+          req.headers['Content-Type'] = 'application/json'
+        end
+        handle_response(response)
+      end
+
+      protected
+
+      def handle_response(response)
```

Comment:
Do we want / have to duplicate this code? There already is response error handling code in so many places of spaceship, if possible we don't want to add _more_ before cleaning that up.

Reply:
I think having this duplication is fine. This is because each client talks to a different host. As such, each host has _slightly_ different response, and error mechanics. It makes sense to be able to tweak those on a per-client basis.